contracts/eosio.distribute/src/eosio.distribute.cpp

```diff
@@ -0,0 +1,95 @@
+#include <eosio.distribute/eosio.distribute.hpp>
+#include <eosio.token/eosio.token.hpp>
+
+namespace eosiodistribute {
+    using eosio::token;
+
+    distribute_contract::distribute_contract( name s, name code, datastream<const char*> ds )
+    :contract(s,code,ds),
+    _distrib_singleton(get_self(), get_self().value),
+    _claimers(get_self(), get_self().value)
+    {
+        _distrib_state = _distrib_singleton.exists() ? _distrib_singleton.get() : distribute_state{};
+    }
+    
+    void distribute_contract::donate_to_rex(const asset& amount, const std::string& memo) {
+        system_contract::donatetorex_action donate_action{ system_account, { get_self(), system_contract::active_permission } };
+        donate_action.send( get_self(), amount, memo );
+    }
+
+    void distribute_contract::setdistrib(const std::vector<distribute_account>& accounts) {
+        require_auth(get_self());
+        
+        if( accounts.size() > 0 ){
+            int64_t remaining_pct = max_distribute_pct;
+            for( distribute_account acct : accounts ) {
+                check(acct.account != get_self(), "Cannot set account to " + get_self().to_string() );
+                check(is_account(acct.account), "Account does not exist: " + acct.account.to_string() );
+                check(0 < acct.percent, "Only positive percentages are allowed");
+                check(acct.percent <= remaining_pct, "Total percentage exceeds 100%");
+                remaining_pct -= acct.percent;
+                if(acct.account == rex_account)
+                    donate_to_rex(asset(0, system_contract::get_core_symbol()), "rex enabled check" );
+            }
+            check(remaining_pct == 0, "Total percentage does not equal 100%");
+        }
+        // set accounts
+        auto& accts = _distrib_state.accounts;
+        accts = accounts;
+
+        _distrib_singleton.set( _distrib_state, get_self() );
+   }
+
+    void distribute_contract::claimdistrib(const name& claimer) {
+        require_auth(claimer);
+
+        auto citr = _claimers.find(claimer.value);
+        check(citr != _claimers.end(), "Not a valid distribution account");
+        if(citr->balance.amount != 0) {
+            token::transfer_action transfer_act{ system_contract::token_account, { get_self(), system_contract::active_permission } };
+            transfer_act.send( get_self(), claimer, citr->balance, "distribution claim" );
+        }
+        _claimers.erase(citr);
+    }
+
+   void distribute_contract::distribute(name from, name to, asset quantity, eosio::ignore<std::string> memo) {
+       if (to != get_self())
+           return;
+        check(quantity.symbol == eosiosystem::system_contract::get_core_symbol(), "Invalid symbol");
+        check(_distrib_state.accounts.size() > 0, "distribution accounts were not setup");
+
+        if(quantity.amount > 0) {
+            asset remaining = quantity;
```

Comment:
Do we want to do anything with the remainder, like rolling it over to the next distribution?